torch/csrc/distributed/autograd/engine/dist_engine.cpp

```diff
@@ -64,6 +65,8 @@ void DistEngine::validateRootsAndRetrieveEdges(
       rootEdges, grads, [](const std::string& msg) { return msg; });
 }
 
+// NB: this function modifies the autograd graph. More specifically it replaces
+// all AccumulateGrad nodes by DistAccumulateGrad.
```

Comment:
> it seems to be confusing that we are replacing nodes in a function called computeDependencies. It might be better to write a new function to do node replacement. We saved some time by avoiding one more traversal. But now the code is much harder to understand and maintain.

I thought about another using another function to traverse the graph again. My consideration is for performance.

Reply:
@albanD @mrshenli What do you guys think about another graph traversal?

Reply:
The vanilla engine does two when using autograd.grad.
The good thing about splitting them is that you will be able to use the `compute_dependencies` function from the original engine and that reduces code duplication.